File: packages/core/src/protocol.ts
export type NativeProps = Record<string, unknown>

export type Operation =
  /** `parent` present means the node is placed by this operation; null is the root list. */
  | { op: "create"; id: number; component: string; props: NativeProps; subscription: number | null; parent?: number | null; before?: number }
  | { op: "props"; id: number; component: string; props: NativeProps }
  | { op: "listen"; id: number; subscription: number | null }
  | { op: "place"; parent: number | null; child: number; before: number | null }
  | { op: "remove"; id: number }
  | { op: "hidden"; id: number; hidden: boolean }
  | { op: "command" | "query"; id: number; component: string; request: number; value: unknown }
  /** A style definition, referenced by later props as `style: id`. */
  | { op: "style"; id: number; style: NativeProps }
  | { op: "dropStyle"; id: number }

export interface Transaction { version: 1; sequence: number; operations: Operation[] }

export interface TransactionReply {
  sequence: number
  /** Retire callbacks only after preceding events have reached the receiver. */
  retired: number[]
  results: Array<{ request: number; value?: unknown; error?: string }>
}

export interface NativeEvent { target: number; subscription: number; payload: unknown }

/** How one prop travels on the binary wire. */
export type WireType = "bool" | "i32" | "u32" | "f32" | "f64" | "str" | "style" | "value"
export interface WireField { name: string; type: WireType; required: boolean }
/** One native component as the worker sees it. Index order is the wire's kind order.
 * `fields` is null when the props type has no positional schema and travels as a map. */
export interface KindSchema {
  name: string
  capabilities: { events: boolean; commands: boolean; queries: boolean; children: boolean; view: boolean }
  fields: WireField[] | null
}

/** A transport is bound to one root/session. It must preserve transaction and
 * event/ack ordering. send resolves after native application, not presentation.
 * The payload is encoded once here and decoded once at native admission. */
export interface Transport {
  /** JSON text or the binary wire. A binary payload is valid until `send` settles. */
  send(transaction: string | Uint8Array): Promise<TransactionReply>
  subscribe(receiver: (event: NativeEvent) => void, onError?: (error: Error) => void): () => void
  close(reason: string): void
}

export interface NativeRef<Command = unknown, Query = unknown, Reply = unknown> {
  readonly id: number
  command(value: Command): Promise<void>
  query(value: Query): Promise<Reply>
}

/** Native draw metadata. This is not an OS presentation acknowledgement. */
export interface FrameInfo {
  root: string
  frame: number
  commit: number
  viewportWidth: number
  viewportHeight: number
  scaleFactor: number
}

File: crates/gpui-react/src/protocol.rs
//! The wire format contains committed changes, never a second retained tree.
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Transaction {
    pub version: u32,
    pub sequence: u64,
    pub operations: Vec<Operation>,
}

#[derive(Debug, Deserialize)]
#[serde(tag = "op", rename_all = "camelCase", deny_unknown_fields)]
pub enum Operation {
    Create {
        id: u64,
        component: String,
        props: Value,
        subscription: Option<u64>,
    },
    Props {
        id: u64,
        props: Value,
    },
    Listen {
        id: u64,
        subscription: Option<u64>,
    },
    Place {
        parent: Option<u64>,
        child: u64,
        before: Option<u64>,
    },
    Remove {
        id: u64,
    },
    Hidden {
        id: u64,
        hidden: bool,
    },
    Command {
        id: u64,
        request: u64,
        value: Value,
    },
    Query {
        id: u64,
        request: u64,
        value: Value,
    },
}

#[derive(Debug, Serialize)]
pub struct Reply {
    pub sequence: u64,
    pub retired: Vec<u64>,
    pub results: Vec<CallResult>,
}

#[derive(Debug, Serialize)]
pub struct CallResult {
    pub request: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

impl CallResult {
    pub(crate) fn new(request: u64, result: anyhow::Result<Value>) -> Self {
        match result {
            Ok(value) => Self {
                request,
                value: Some(value),
                error: None,
            },
            Err(error) => Self {
                request,
                value: None,
                error: Some(format!("{error:#}")),
            },
        }
    }
}
